### PR TITLE
Update categorical.py

### PR DESCRIPTION
**Pull request by harrismirza**
Added change to Categorical Entropy to ignore non-finite probability values.

### CODE REVIEW

**Comment by harrismirza** [comment]
CLA is done

File: tensorflow_probability/python/distributions/categorical.py [hunk 1]
```diff
@@ -299,8 +299,8 @@ def _log_prob(self, k):
                                                            logits=logits)
 
   def _entropy(self):
-    return -tf.reduce_sum(
-        tf.nn.log_softmax(self.logits) * self.probs, axis=-1)
+    entropies = self.probs * tf.log(self.probs)
+    return -tf.reduce_sum(tf.boolean_mask(entropies, tf.is_finite(entropies)), axis=-1)
```
Comment (langmore) [inline tensorflow_probability/python/distributions/categorical.py]: Thank you for the contribution.  I see two issues here.

First, I see how this happens when probs is zero and someone does `tf.log(self.probs)`.  However, `tf.nn.log_softmax(self.logits)` is different.  What am I missing?

Second, note that `tf.boolean_mask(entropies, tf.is_finite(entropies))` will return a rank 1 Tensor, and then reduction over axis = -1 will return a scalar.  This is not what you want...you need to reduce only over the last dimension.  I guess the tests were not complete enough to catch this.

Comment (harrismirza) [reply tensorflow_probability/python/distributions/categorical.py]: Sorry about the commit being wrong, I'm quite new to Tensorflow.

Log Softmax is implemented as `logits - log(reduce_sum(exp(logits), axis))` according to the API docs, and when a Categorical Distribution is created with the probabilities `[0, 0.5, 0.5]`, it stores the logits as `[-inf, -0.69..., -0.69...]`.

So when calculating Log Softmax for the first value it calculates:
= `-inf - log(e^-inf + e^-0.69 + e^0.69)
= -inf - log(0 + 0.5 + 0.5)
= -inf - log(1)
= -inf - 0
= -inf
`
And when this is multiplied by the probability of 0, python evaluates -inf * 0 as -inf.

Comment (axch) [reply tensorflow_probability/python/distributions/categorical.py]: I suggest just computing the entropy terms with `tf.xlogy` and dispensing with the mask, like this:

```suggestion
    return -tf.reduce_sum(tf.xlogy(self.probs, self.probs), axis=-1)
```

`xlogy` is designed for exactly this: by definition, `xlogy(0, 0) == 0`, not `NaN`.

For us to approve the pull request, it would also be great if you could add a unit-test of entropy in the presence of zeros that fails before the change and passes after.

Thanks again for your contribution!

File: tensorflow_probability/python/distributions/categorical.py [hunk 2]
```diff
@@ -299,8 +299,8 @@ def _log_prob(self, k):
                                                            logits=logits)
 
   def _entropy(self):
-    return -tf.reduce_sum(
-        tf.nn.log_softmax(self.logits) * self.probs, axis=-1)
+    entropies = self.probs * tf.log(self.probs)
```
Comment (brianwa84) [inline tensorflow_probability/python/distributions/categorical.py]: entropies looks unused

Comment (ebrevdo) [reply tensorflow_probability/python/distributions/categorical.py]: remove this line.

File: tensorflow_probability/python/distributions/categorical.py [hunk 3]
```diff
@@ -299,8 +299,8 @@ def _log_prob(self, k):
...
+    return -tf.reduce_sum(tf.xlogy(self.probs, self.probs), axis=-1)
```
Comment (ebrevdo) [inline tensorflow_probability/python/distributions/categorical.py]: can you change this to:

```
return -tf.reduce_sum(tf.math.multiply_no_nan(tf.nn.log_softmax(self.logits), self.probs), axis=-1)
```

thanks!
